### PR TITLE
Add dynamic config to disable loading task queue user data

### LINKED ISSUE #4450: Add dynamic config to disable loading task queue user data
Task queue user data is now in the critical path and loaded for every task queue partition (only the root reads from the DB).
At the time of writing only versioning data is stored in user data and until we're confident with the user data mechanism, we need the ability to disable it.

### PR DESCRIPTION
**Pull request by bergundy**
**What changed?**

See PR title

**Why?**

Closes #4450 
User data is a new feature and effects every task queue load in matching. We need a way to disable it.

**How did you test it?**

Unit tests and functional tests.

### CODE REVIEW

**Review by yiminc — APPROVED** [review]

File: service/matching/taskQueueManager.go [hunk 1]
```diff
@@ -517,11 +519,17 @@ func (c *taskQueueManagerImpl) GetUserData(ctx context.Context) (*persistencespb
 	if c.managesSpecificVersionSet() {
 		return nil, nil, errNoUserDataOnVersionedTQM
 	}
+	if !c.config.LoadUserData() {
+		return nil, nil, nil
```
Comment (dnr) [inline service/matching/taskQueueManager.go:523]: Did you consider having this function return a sentinel error in this case and having the other "disable versioning" stuff look at that error instead of separately evaluating `config.LoadUserData()`? At first glance, I think that would be cleaner since the config is only being used in one place

File: tests/versioning_test.go [hunk 2]
```diff
@@ -39,14 +39,14 @@ import (
 
 	enumspb "go.temporal.io/api/enums/v1"
 	"go.temporal.io/api/serviceerror"
-	"go.temporal.io/api/workflowservice/v1"
+	workflowservice "go.temporal.io/api/workflowservice/v1"
```
Comment (dnr) [inline tests/versioning_test.go:42]: undo this? looks like an over-aggressive goimports

File: tests/versioning_test.go [hunk 3]
```diff
@@ -1449,6 +1565,15 @@ func (s *versioningIntegSuite) waitForChan(ctx context.Context, ch chan struct{}
 	}
 }
 
+func (s *versioningIntegSuite) unloadTaskQueue(ctx context.Context, tq string) {
```
Comment (dnr) [inline tests/versioning_test.go:1568]: change line 179 to use the new helper

**Comment by dnr** [comment]
Also, this doesn't disable the call to `store.GetTaskQueueUserData` in db. We should disable that one too to disable more code paths that could fail (e.g. imagine we had a bug in deserialization)

File: service/matching/config.go [hunk 4]
```diff
@@ -206,6 +214,9 @@ func newTaskQueueConfig(id *taskQueueID, config *Config, namespace namespace.Nam
 			return config.SyncMatchWaitDuration(namespace.String(), taskQueueName, taskType)
 		},
 		TestDisableSyncMatch: config.TestDisableSyncMatch,
+		LoadUserData: func() bool {
+			return config.LoadUserData(namespace.String(), taskQueueName, taskType)
```
[the remark this answers is not in the document]
Comment (bergundy) [reply service/matching/config.go:218]: Yeah, I don't expect anyone to specify an override at this granularity.

File: service/matching/taskQueueManager.go [hunk 5]
```diff
@@ -315,6 +315,8 @@ func (c *taskQueueManagerImpl) Start() {
 	c.taskReader.Start()
 	if c.shouldFetchUserData() {
 		c.goroGroup.Go(c.fetchUserDataLoop)
+	} else {
+		c.userDataInitialFetch.Set(struct{}{}, nil)
```
[the remark this answers is not in the document]
Comment (bergundy) [reply service/matching/taskQueueManager.go:319]: I had an edge case where the future.Set call would panic in `fetchUserDataLoop` when I changed the dynamic config, that's why I moved it here.
Feel free to find a better place for it.

FTR the test that sometimes repros the panic is TestWorkflowQueryTimesOutWhenDisablingLoadingUserData.

File: service/matching/taskQueueManager.go (hunk 1 above)
Comment (bergundy) [reply service/matching/taskQueueManager.go:523]: I like that this config is fairly contained in the code and doesn't leak all over.
Remember that this is a temporary flag (or so we hope) and I want to be able to easily remove it later.

File: tests/versioning_test.go (hunk 2 above)
Comment (bergundy) [reply tests/versioning_test.go:42]: The linter is not smart enough to know that if you have a variable named `v1` and an import line that ends with `v1` there's no shadowing.

**Comment by bergundy** [comment]
> Also, this doesn't disable the call to `store.GetTaskQueueUserData` in db. We should disable that one too to disable more code paths that could fail (e.g. imagine we had a bug in deserialization)

But we disable this in `taskQueueManagerImpl.GetUserData`, which is where the db call is initiated.
Are you suggesting we move the dynamic config lookup into db?

File: service/matching/taskQueueManager.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (dnr) [reply service/matching/taskQueueManager.go:319]: I moved this around in https://github.com/temporalio/temporal/pull/4487, I think it's better. See what you think

File: service/matching/taskQueueManager.go (hunk 1 above)
Comment (dnr) [reply service/matching/taskQueueManager.go:523]: I tried this refactoring and I like it so far.. I'll send another PR later for discussion

File: tests/versioning_test.go (hunk 2 above)
Comment (dnr) [reply tests/versioning_test.go:42]: I'd rather fix/disable the broken linter than appease it
